hivemind/averaging/control.py

```diff
@@ -0,0 +1,136 @@
+import struct
+from enum import Enum
+from typing import Optional
+
+import numpy as np
+import torch
+
+from hivemind.utils import DHTExpiration, MPFuture, get_dht_time, get_logger
+
+logger = get_logger(__name__)
+
+
+class AveragingStage(Enum):
+    IDLE = 0  # still initializing
+    LOOKING_FOR_GROUP = 1  # running decentralized matchmaking, can't run allreduce yet
+    AWAITING_TRIGGER = 2  # waiting for user to set the trigger that allows running allreduce
+    RUNNING_ALLREDUCE = 3  # exchanging tensors with groupmates
+    FINISHED = 4  # either done or failed with exception
+
+
+class StepControl(MPFuture):
+    """
+    An auxiliary data structure that allows user to control stages and track progress in a single averaging step
+
+    :param scheduled_time: estimated time when averaging should begin. Will be used for scheduling
+    :param deadline: if averaging is still in progress at this time, it should be stopped due to TimeoutError
+    :param allow_retries: if True, allow running matchmaking and all-reduce again if previous attempt fails
+    :param weight: averaging weight, can be changed afterwards
+    :param gather_binary: optionally send this data to all peers in the next group and gather it from groupmates
+    :returns: an assembled group if successful, None if failed; does NOT perform the actual averaging
+    """
+
+    def __init__(
+        self, scheduled_time: DHTExpiration, deadline: float, allow_retries: bool, weight: float, gather_binary: bytes
+    ):
+        super().__init__()
+        self._gather_binary, self._deadline, self._allow_retries = gather_binary, deadline, allow_retries
+        self._trigger: Optional[MPFuture] = None
+
+        # Buffer contents:
+        # scheduled_time (double) | weight (double) | stage (AveragingStage, 1 byte) | began_allreduce: (bool, 1 byte)
+        self._shared_buffer = torch.zeros([18], dtype=torch.uint8).share_memory_()
+        self.stage = AveragingStage.IDLE
+        self.scheduled_time = scheduled_time
+        self.weight = weight
+        self.began_allreduce = False
+
+    def attach_trigger(self, trigger: MPFuture):
+        assert self._trigger is None, "trigger is already attached"
+        self._trigger = trigger
+
+    def allow_allreduce(self):
+        """Allow averager to begin allreduce when it finds a group. Meant to be triggered by user."""
+        assert self._trigger is not None, "StepControl does not have an attached trigger (not properly initialized)"
+        if self._trigger.done():
+            logger.warning("Trigger is already set")
+        else:
+            self._trigger.set_result(None)
+
+    async def wait_for_trigger(self):
+        assert self._trigger is not None, "StepControl does not have an attached trigger (not properly initialized)"
+        await self._trigger
+
+    @property
+    def scheduled_time(self) -> DHTExpiration:
+        return struct.unpack("d", self._shared_buffer[0:8].numpy().data)[0]
```

Comment:
Is it possible to just use four `mp.Value`s instead of playing around with byte representations? What are the constraints for _shared_buffer here that necessitate the shared-array part?

Furthermore, maybe pytorch isn't necessary and something like [SharedMemory](https://docs.python.org/3/library/multiprocessing.shared_memory.html#multiprocessing.shared_memory.SharedMemory) can help us do the same task with less levels of abstraction

Reply:
If having this array is absolutely necessary, I'd create a dict/namedtuple of slices for its different parts to avoid indexing bugs

Reply:
MP Value cannot be passed ahead without inheritance
shared memory is incompatible with py37
if you have any ideas, please apply them to MPFuture first

Reply:
What about four shared pytorch tensors then? My only concern is that stacking all values in a single array might be the easiest solution storage-wise, but more difficult to understand, maintain and wrap one's head around of

Reply:
Four pytorch tensors is very expensive w.r.t. the number of open files, that's the same reason we have byte shifts in MPFuture